File: src/udemy_tutorials_py/udemy_tutorials_py/number_counter.py
#!/usr/bin/env python3
"""Python node that takes a number and counts from 0 to that number.

    Publishers:
        number_count (std_msgs/Int64) - Publishes the current count.

    Subscribers:
        number (std_msgs/Int64) - Subscribes to get the value that needs to be
        added to the counter.
"""
import sys
import rclpy
from rclpy.node import Node
from std_msgs.msg import Int64
from example_interfaces.srv import SetBool


class NumberCount(Node):
    """Node that counts from 0 by adding a number to the count."""
    def __init__(self):
        super().__init__("number_counter")
        self.__count = 0
        self.subscriber = self.create_subscription(
            msg_type=Int64,
            topic="number",
            callback=self.callback_number_received,
            qos_profile=10,
        )
        self.publisher = self.create_publisher(
            msg_type=Int64,
            topic="number_count",
            qos_profile=10,
        )
        self.reset_counter_service = self.create_service(
            srv_type=SetBool,
            srv_name="reset_counter",
            callback=self.callback_reset_counter,
        )
        self.logger.info("Number Counter has been started.")
        self.logger.info(f"Counter: {self.__count}")

    @property
    def logger(self):
        """Logger for the node."""
        return self.get_logger()

    def callback_number_received(self, msg):
        """Callback function that is called when a number is received."""
        self.__count += msg.data
        self.logger.info(
            f"Received number: {msg.data}. Current count: {self.__count}")
        self.publish_count()

    def publish_count(self):
        """Publishes the current count."""
        msg = Int64()
        msg.data = self.__count
        self.publisher.publish(msg)
        self.logger.info(f"Published count: {msg.data}")

    def callback_reset_counter(self, request, response):
        """Callback function for the reset counter service."""
        current_value = self.__count
        if request.data:
            self.__count = 0
            self.logger.info("Counter reset.!!!")
        response.success = True
        response.message = f"Counter reset. Old Value: {current_value}"
        return response


def main(args=None):
    """Main function that creates the node and spins it."""
    rclpy.init(args=args)
    node = NumberCount()
    rclpy.spin(node)
    rclpy.shutdown()
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/udemy_tutorials_cpp/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/int64.hpp"
#include "example_interfaces/srv/set_bool.hpp"

using std::placeholders::_1;
using std::placeholders::_2;

class NumberCounter : public rclcpp::Node
{
    public:
        NumberCounter(): Node("number_counter")
        {
            // Create subscriber to listen to the topic "number"
            // The callback function is called when a message is received
            // The message is of type std_msgs::msg::Int64
            subscriber_ = this->create_subscription<std_msgs::msg::Int64>(
                "number",
                10,  // Queue size
                std::bind(&NumberCounter::number_received,
                          this,
                          std::placeholders::_1)
            );
            RCLCPP_INFO(this->get_logger(), "Number counter has been started.");
            publisher_ = this->create_publisher<std_msgs::msg::Int64>(
                "number_count",
                10
            );
            service_ = this->create_service<example_interfaces::srv::SetBool>(
                "reset_counter",
                std::bind(&NumberCounter::callback_reset_counter, this, _1, _2)
            );
        } 

    private:
        void number_received(const std_msgs::msg::Int64::SharedPtr msg)
        {
            // Increment the counter
            counter_ += msg->data;
            RCLCPP_INFO(this->get_logger(), "Counter: %d", counter_);
            this->publish_counter();
        }

        void publish_counter()
        {
            // Publish the counter value
            auto message = std_msgs::msg::Int64();
            message.data = counter_;
            publisher_->publish(message);
            RCLCPP_INFO(this->get_logger(),
                        "Counter value has been published (from function).");
        }

        void callback_reset_counter(const example_interfaces::srv::SetBool::Request::SharedPtr request,
                                    const example_interfaces::srv::SetBool::Response::SharedPtr response)
        {
            if (request->data)
            {
                counter_ = 0;
                RCLCPP_INFO(this->get_logger(), "Counter has been reset.");
                response->success = true;
                response->message = "Counter has been reset.";
            }
            else
            {
                response->success = false;
                response->message = "Counter has not been reset.";
            }
        }
        int counter_ = 0;
        rclcpp::Subscription<std_msgs::msg::Int64>::SharedPtr subscriber_;
        rclcpp::Publisher<std_msgs::msg::Int64>::SharedPtr publisher_;
        rclcpp::Service<example_interfaces::srv::SetBool>::SharedPtr service_;
};


int main(int argc, char **argv)
{
    // Initialize ROS2
    rclcpp::init(argc, argv);
    // Create a NumberCounter node
    auto node = std::make_shared<NumberCounter>();
    // Spin the node
    rclcpp::spin(node);
    // Shut down ROS2
    rclcpp::shutdown();
    return 0;
}
